aten/src/ATen/quantized/Quantizer.cpp

```diff
@@ -11,11 +11,112 @@
 
 namespace at {
 
+#ifdef USE_FBGEMM
+template <typename T>
+T quantize_val(float scale, int32_t zero_point, float value) {
+  // Internally, fbgemm::Quantize uses std::nearbyint.
+  // std::nearbyint results in nearest integer value according to the current
+  // rounding mode and the default rounding mode is rounds to even in half-way
+  // cases in most popular processor architectures like x86 and ARM. This is
+  // typically faster than an alternatives like std::round that rounds half-way
+  // cases away from zero, and can be consistent with SIMD implementations for
+  // example in x86 using _mm512_cvtps_epi32 or mm512_round_ps with
+  // _MM_FROUND_CUR_DIRECTION option that also follow the current rounding mode.
+  int32_t qvalue;
+  qvalue = fbgemm::Quantize<typename T::underlying>(value, zero_point, scale,
+                                                    /*result_precision=*/std::numeric_limits<typename T::underlying>::digits);
+  return static_cast<T>(qvalue);
+}
+
+template <typename T>
+Tensor quantize_tensor(Tensor rtensor, Tensor qtensor, float scale, int32_t zero_point) {
+  const float* rd = rtensor.data<float>();
+  auto qd = reinterpret_cast<typename T::underlying*>(qtensor.data<T>());
+  fbgemm::TensorQuantizationParams qparams;
+  qparams.scale = scale;
+  qparams.zero_point = zero_point;
+  qparams.precision = std::numeric_limits<typename T::underlying>::digits;
+  fbgemm::Quantize<typename T::underlying>(/*src=*/rd,
+                             /*dst=*/qd,
+                             /*len=*/rtensor.numel(),
+                             /*qparams=*/qparams);
+  return qtensor;
+}
+
+template <typename T>
+Tensor dequantize_tensor(Tensor qtensor, Tensor rtensor, float scale, int32_t zero_point) {
+  const auto* qd = reinterpret_cast<const typename T::underlying*>(qtensor.data<T>());
+  fbgemm::TensorQuantizationParams qparams;
+  qparams.scale = scale;
+  qparams.zero_point = zero_point;
+  qparams.precision = std::numeric_limits<typename T::underlying>::digits;
+  float* rd = rtensor.data<float>();
```

Comment:
is this the only thing we need to check? or should we also check for type of qtensor matching template T as well etc.

I think one question is whether we want to expose this function to user, if not we only need to check for float since the qtensor is constructed by us.

Reply:
yes, you should check that to.

as for exposed to user vs not, if you are taking in a parameter, you should check it.  What differs is if you use `AT_CHECK` or `AT_ASSERT` for external vs internal errors.